python/astshim/fitsChanContinued.py

```diff
@@ -0,0 +1,258 @@
+                else:
+                    raise RuntimeError(f"Type, {ctype} of FITS card '{name}' not supported")
+
+                # Increment the card number to continue search
+                self.setCard(thiscard + 1)
+        finally:
+            # Reinstate the original card position
+            self.setCard(currentCard)
+
+        if not values:
+            raise KeyError(f"{name}")
+
+        # We may have multiple values. Unlike pyast we always return a tuple.
+        return tuple(values)
+
+    raise ValueError(f"Supplied key, '{name}' of unsupported type")
+
+
+FitsChan.__getitem__ = getitem
+
+
+def setitem(self, name, value):
+    """name can be integer index or keyword.
+
+    If an integer, the value is deemed to be a full card to replace the
+    existing one. If the value is None or empty string a blank header
+    card is created at the location.
+
+    This can affect the position of the current card since cards can be
+    inserted."""
+
+    if isinstance(name, int):
+        # Calculate position in FitsChan (0-based to 1-based)
+        newpos = _calc_card_pos(self, name)
+        self.setCard(newpos)
+
+        if not value:
+            value = " "
+
+        # Overwrite the entire value
+        self.putFits(value, True)
+        return
+
+    # A blank name results in a comment card being inserted at the
+    # current position
+    if not name:
+        self.setFitsCM(value, False)
+        return
+
+    # Get current card position and rewind
+    currentCard = self.getCard()
+    try:
+        self.clearCard()
+
+        # Look for a card with the specified name
+        # We do not care about the result, if nothing is found we will be at the
+        # end of the header
+        self.findFits(name, False)
+
+        # pyast seems to want to delete items if the value is None but
+        # astropy and PropertyList think the key should be undefined.
+        if value is None:
+            self.setFitsU(name, overwrite=True)
+        elif isinstance(value, int):
+            self.setFitsI(name, value, overwrite=True)
+        elif isinstance(value, float):
+            self.setFitsF(name, value, overwrite=True)
+        elif isinstance(value, bool):
+            self.setFitsL(name, value, overwrite=True)
+        else:
+            # Force to string
+            self.setFitsS(name, str(value), overwrite=True)
+
+        # Delete any later cards with matching keyword
+        while True:
+            found = self.findFits(name, False)
+            if not found.found:
+                break
+            self.delFits()
+    finally:
+        # Try to reinstate the current card
+        self.setCard(currentCard)
+
+
+FitsChan.__setitem__ = setitem
+
+
+def delitem(self, name):
+    """Delete an item from the FitsChan either by index (0-based) or by name.
+
+    If a name is given all instances of the name will be deleted.
+    The current card position is not retained.
```

Comment:
This is great. One minor request: please document what happens if name is an integer and is out of bounds. I hope it raises an exception, but which one?

Also please clarify what happens if name is a string and does not match any cards. I am guessing it silently does nothing, but please say so.

Reply:
`del` should probably raise an exception if the requested key is not found (as happens in a normal dict).